File: algs-sandbox-python/primitivesAndBinary/power_of_two.py


"""
Given a non-negative integer, return true if the number is a power of two, return false otherwise.

"""

def power_of_two(x): 
    # 8
    # 00001000
    # 6
    # 00000110
    # powers of two are where only 1 bit is set
    num_bits_set = 0
    while x > 0:
        if (x & 1):
            num_bits_set += 1
        x >> 1
    return num_bits_set == 1



assert power_of_two(1) == True
assert power_of_two(8) == True
assert power_of_two(0) == False
assert power_of_two(0) == False

File: algs-sandbox-java/src/primitives/PowerOfTwo.java
package primitives;

public class PowerOfTwo {
    public static void main(String[] args) {
        System.out.println("Is 16 a power of two? " + isPowerOfTwo(16));
        System.out.println("Is 7 a power of two? " + isPowerOfTwo(7));
        System.out.println("Is 32 a power of two? " + isPowerOfTwo(32));
        System.out.println("Is 36 a power of two? " + isPowerOfTwo(36));
    }

    public static boolean isPowerOfTwo(int n) { 
        // check if number and'd with bit mask returns exactly zero
        if (n > 0 && (n & (n - 1)) == 0) {
            return true;
        }
        return false;
    }
}
